sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -3232,6 +3232,35 @@ def compile_and_reload(
             pipeline_func=pipeline, package_path=output_yaml)
         return components.load_component_from_file(output_yaml)
 
+class TestResourceConfig(unittest.TestCase):
+    def test_cpu_memory_optional(self):
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            predict_op = components.load_component_from_text("""
+      name: predict
+      inputs:
+      - {name: generate_explanation, type: Boolean, default: False}
+      implementation:
+        container:
+          image: gcr.io/my-project/my-image:tag
+          args:
+          - {inputValue: generate_explanation}
+      """)
+
+            @dsl.pipeline(name='test-resource-pipeline')
+            def simple_pipeline():
+                predict_op(generate_explanation=True)
+                predict_op(generate_explanation=True).set_cpu_limit('5')
+                predict_op(generate_explanation=True).set_memory_limit('50G')
+                predict_op(generate_explanation=True).set_cpu_limit('5').set_memory_limit('50G')
+
+            target_json_file = os.path.join(tmpdir, 'result.yaml')
+            compiler.Compiler().compile(
+                pipeline_func=simple_pipeline, package_path=target_json_file)
```

Comment:
Instead of compiling, we can just make assertions on the `simple_pipeline.pipeline_spec` message itself, which invokes compilation under the hood.

Reply:
accessing child field with `simple_pipeline.pipeline_spec` is cumbersome, so I convert it to `dict` first. Hope it works. 

All other comments are addressed.